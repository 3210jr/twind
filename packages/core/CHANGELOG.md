# @twind/core

## 1.0.0-next.1

### Patch Changes

- Initial publish of twind v1 preview ([`179b9653`](https://github.com/tw-in-js/twind/commit/179b9653de661a62a661c80d5506ae68f7964aba))
