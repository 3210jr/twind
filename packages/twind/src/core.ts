/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-explicit-any */

// Fix exported global variable from `twindCdn` to `twind`
/* @distilt-global-name twind */

import type {
  Twind,
  BaseTheme,
  TwindConfig,
  TwindUserConfig,
  Preset,
  ExtractThemes,
  Sheet,
} from '@twind/core'

import { init, autoInit } from '@twind/runtime'

export * from '@twind/core'
export { tw } from '@twind/runtime'

const cancelAutoInit = autoInit(setup)

export function setup<Theme extends BaseTheme = BaseTheme, SheetTarget = CSSStyleSheet | string[]>(
  config?: TwindConfig<Theme>,
  sheet?: Sheet<SheetTarget>,
  target?: HTMLElement,
): Twind<Theme, SheetTarget>

export function setup<
  Theme = BaseTheme,
  Presets extends Preset<any>[] = Preset[],
  SheetTarget = CSSStyleSheet | string[],
>(
  config?: TwindUserConfig<Theme, Presets>,
  sheet?: Sheet<SheetTarget>,
  target?: HTMLElement,
): Twind<BaseTheme & ExtractThemes<Theme, Presets>, SheetTarget>

export function setup(
  config: TwindConfig<any> | TwindUserConfig<any> = {},
  sheet?: Sheet,
  target?: HTMLElement,
): Twind {
  cancelAutoInit()

  return init(config as TwindConfig<any>, sheet, target)
}
