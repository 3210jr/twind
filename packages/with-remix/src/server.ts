/* eslint-env node */

export { inline as default } from 'twind'

export type { InlineOptions, InlineMinify } from 'twind'
