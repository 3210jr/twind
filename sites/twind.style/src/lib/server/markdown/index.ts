import * as path from 'node:path'
import fg, { type Options as FastGlobOptions } from 'fast-glob'
import { findUpSync, pathExistsSync } from 'find-up'
import grayMatter from 'gray-matter'
const { read: readFile } = grayMatter

export interface Page {
  path: string
  href: string
  section: string | undefined
  label: string | undefined
  title: string | undefined
  excerpt: string | undefined
  file: string
  prev?: string | undefined
  next: string | undefined
}

export interface ReadOptions extends FastGlobOptions {
  slug?: string
  root?: string
  pattern?: string | string[]
}

export function read({
  slug = '/',
  cwd = process.cwd(),
  root = findUpSync(
    (directory) =>
      ['.git', 'pnpm-workspace.yaml', 'lerna.json'].some((file) =>
        pathExistsSync(path.join(directory, file)),
      )
        ? directory
        : undefined,
    { cwd, type: 'directory' },
  ) || cwd,
  pattern = '**/*.md',
  ignore = ['**/README.md'],
  caseSensitiveMatch = false,
  ...globOptions
}: ReadOptions = {}) {
  const pages = new Map<string | undefined, Page>()

  for (const filename of fg.sync(pattern, { ...globOptions, cwd, ignore, absolute: true })) {
    const href = toHref(cwd, filename, slug)

    const { data: frontmatter } = readFile(filename)

    const { section, label, title, excerpt, next } = frontmatter

    pages.set(href, {
      path: filename,
      href,
      section,
      label: label || title || path.basename(filename),
      title: title || label || path.basename(filename),
      excerpt,
      file: path.relative(root, filename),
      next: next && toHref(cwd, path.resolve(path.dirname(filename), next), slug),
    })
  }

  // update prev link
  pages.forEach((value, href) => {
    const next = pages.get(value.next)

    if (next) {
      next.prev = href
    }
  })

  let startHref: string | undefined = undefined
  for (const [href, entry] of pages) {
    if (entry.next && !entry.prev) {
      startHref = href
      break
    }
  }

  if (!startHref) startHref = pages.keys().next().value

  const seen = new Set<Page>()
  const sections = new Map<string | undefined, string[]>([[undefined, []]])
  for (let current = pages.get(startHref); current; current = pages.get(current.next)) {
    seen.add(current)

    let section = sections.get(current.section)
    if (!section) {
      sections.set(current.section, (section = []))
    }

    section.push(current.href)
  }

  // collect those that are not linked
  for (const current of pages.values()) {
    if (seen.has(current)) continue

    let section = sections.get(current.section)
    if (!section) {
      sections.set(current.section, (section = []))
    }

    section.push(current.href)
  }

  return { startHref, pages, sections }
}

function toHref(cwd: string, file: string, slug = '/') {
  return slug + path.relative(cwd, omitExtension(file))
}

function omitExtension(file: string) {
  return file.replace(/\.[^.]+$/, '')
}
