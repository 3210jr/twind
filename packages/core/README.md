# @twind/core

Minimal implementation of a tailwind-compatible CSS-in-JS framework.

**This package does not contain any Tailwindcss rules. These are defined in [@twind/preset-tailwind](../preset-tailwind).**

---

## READ THIS FIRST!

**Twind v1 is still in beta. Expect bugs!**

---

## Installation

Install from npm:

```sh
# Using npm
npm install @twind/core@next

# Using Yarn
yarn add @twind/core@next
```

## Usage

```js
import { twind, cssom, observe } from '@twind/core'

const tw = observe(
  twind(
    {
      theme: {
        /* .. */
      },
      rules: [
        /* ... */
      ],
    },
    cssom(),
  ),
)
```

## API

### `twind`

TDB

### `cx`

```js
import { cx } from '@twind/core'

// Set a className
element.className = cx`
  underline
  /* multi
    line
    comment
  */
  hover:focus:!{
    sm:{italic why}
    lg:-{px}
    -mx-1
  }
  // Position
  !top-1 !-bottom-2
  text-{xl black}
`
```

### `shortcut`

TDB

### `style`

TDB

### `consume`

Used for static HTML processing (usually to provide SSR support for your javascript-powered web apps)

1. parse the markup and process element classes with the provided Twind instance
2. update the class attributes _if_ necessary
3. return the HTML string with the final element classes

```js
import { twind, virtual, consume} from '@twind/core'

// can be re-used
const tw = twind({ /* config */, virtual()}

function render() {
  const html = app()

  // clear all styles
  tw.clear()

  // generated markup
  const markup = comsume(html, tw)

  // create CSS
  const css = tw.target.join('')

  // inject as last element into the head
  return markup.replace('</head>', `<style id="tw">${css}</style></head>`)
}
```
