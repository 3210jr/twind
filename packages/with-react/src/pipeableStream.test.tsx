import { text } from 'node:stream/consumers'

import { expect, test } from 'vitest'

import { twind, virtual } from '@twind/core'
import presetTailwind from '@twind/preset-tailwind'

import * as React from 'react'
import { renderToPipeableStream } from 'react-dom/server.node'
import TwindStream from './pipeableStream'

function App({ children }: { children?: React.ReactNode }) {
  return (
    <html>
      <head>
        <meta charSet="utf-8" />
      </head>
      <body>{children}</body>
    </html>
  )
}

function HelloWorld() {
  return <h1 className="text-3xl font-bold underline">Hello world!</h1>
}

test('renderToReadableStream', async () => {
  const tw = twind(
    {
      presets: [presetTailwind({ disablePreflight: true })],
    },
    virtual(true),
  )

  let caughtError: unknown | null = null

  const stream = renderToPipeableStream(
    <App>
      <HelloWorld />
    </App>,
    {
      onError(error) {
        caughtError = error
      },
    },
  )

  const body = stream.pipe(new TwindStream(tw))

  const result = await text(body)

  expect(caughtError).toBeNull()
  expect(result).toMatchSnapshot()
})

test('renderToReadableStream with suspense', async () => {
  const tw = twind(
    {
      presets: [presetTailwind({ disablePreflight: true })],
    },
    virtual(true),
  )

  let fullfilled = false
  let promise = null

  const useTimeout = (ms: number) => {
    // check if timeout already occurred.
    if (!fullfilled) {
      // if promise doesn't exist create and throw it.
      throw (promise ||= new Promise((resolve) => {
        setTimeout(() => {
          // on next attempt consider timeout completed.
          fullfilled = true
          // resolve promise (will ask react to re-render).
          resolve(null)
        }, ms)
      }))
    }
  }

  const HowAreYou = () => {
    useTimeout(5)
    return <p className="text-lg font-bold italic">How are you?</p>
  }

  let caughtError: unknown | null = null

  const stream = renderToPipeableStream(
    <App>
      <HelloWorld />
      <React.Suspense>
        <HowAreYou />
      </React.Suspense>
    </App>,
    {
      onError(error) {
        caughtError = error
      },
    },
  )

  const body = stream.pipe(new TwindStream(tw))

  const result = await text(body)

  expect(caughtError).toBeNull()
  expect(result).toMatchSnapshot()
})
