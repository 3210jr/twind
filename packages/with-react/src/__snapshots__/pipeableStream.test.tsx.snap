// Vitest Snapshot v1

exports[`renderToReadableStream 1`] = `"<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><style data-twind>/*!dbgidc,w,text-3xl*/.text-3xl{font-size:1.875rem;line-height:2.25rem}/*!dbgidc,y,font-bold*/.font-bold{font-weight:700}/*!dbgidc,11,underline*/.underline{text-decoration-line:underline}</style></head><body><h1 class=\\"text-3xl font-bold underline\\">Hello world!</h1></body></html>"`;

exports[`renderToReadableStream with suspense 1`] = `"<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><style data-twind>/*!dbgidc,w,text-3xl*/.text-3xl{font-size:1.875rem;line-height:2.25rem}/*!dbgidc,y,font-bold*/.font-bold{font-weight:700}/*!dbgidc,11,underline*/.underline{text-decoration-line:underline}</style></head><body><h1 class=\\"text-3xl font-bold underline\\">Hello world!</h1><!--$?--><template id=\\"B:0\\"></template><!--/$--></body></html><script>!function(e,n){e&&(e.textContent=n.reduce((function(e,n){return e.slice(0,n[0])+n[1]+e.slice(n[0])}),e.textContent||''))}(document.querySelector('style[data-twind=\\"\\"]'),[[81,\\"w,text-lg*/.text-lg{font-size:1.125rem;line-height:1.75rem}/*!dbgidc,\\"],[189,\\"}/*!dbgidc,y,italic*/.italic{font-style:italic\\"]])</script><div hidden id=\\"S:0\\"><p class=\\"text-lg font-bold italic\\">How are you?</p></div><script>function $RC(a,b){a=document.getElementById(a);b=document.getElementById(b);b.parentNode.removeChild(b);if(a){a=a.previousSibling;var f=a.parentNode,c=a.nextSibling,e=0;do{if(c&&8===c.nodeType){var d=c.data;if(\\"/$\\"===d)if(0===e)break;else e--;else\\"$\\"!==d&&\\"$?\\"!==d&&\\"$!\\"!==d||e++}d=c.nextSibling;f.removeChild(c);c=d}while(c);for(;b.firstChild;)f.insertBefore(b.firstChild,c);a.data=\\"$\\";a._reactRetry&&a._reactRetry()}};$RC(\\"B:0\\",\\"S:0\\")</script>"`;
